### PR TITLE
Omezarr v4

### PR DESCRIPTION
**Pull request by KateMoreva**

### CODE REVIEW

**Comment by constantinpape** [comment]
Thanks for working on this @KateMoreva. Is it ready to be merged? @tischi could you have a look? It would be great if we can get this into MoBIE beta until Thursday in order to demo it.

**Comment by KateMoreva** [comment]
> Thanks for working on this @KateMoreva. Is it ready to be merged? @tischi could you have a look? It would be great if we can get this into MoBIE beta until Thursday in order to demo it.

I'll test this a bit more today from MoBIE side as there were some troubles with the transformations.

**Comment by tischi** [comment]
OK, let me know once it's ready.

**Comment by tischi** [comment]
@constantinpape should we merge or wait for the discussions today and tomorrow with the NGFF crowd?

**Comment by constantinpape** [comment]
> @constantinpape should we merge or wait for the discussions today and tomorrow with the NGFF crowd?

Let's merge it and try to deploy on MoBIE beta so that we can do a quick demo tomorrow. If there are some changes to 0.4 they will be small and we can implement them in a follow up PR.

**Comment by tischi** [comment]
Do we have v0.4 example data for the demo?

File: src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java [hunk 1]
```diff
@@ -42,12 +44,41 @@ default String getDimensionSeparator(HashMap<String, JsonElement> attributes) {
...
+           return;
+       }
+       String version = multiscales.getAsJsonArray().get(0).getAsJsonObject().get("version").getAsString();
+        if (version.equals("0.3")) {
             JsonElement axes = multiscales.getAsJsonArray().get(0).getAsJsonObject().get("axes");
             setAxes(axes);
+        } else if (version.equals("0.4")) {
+            JsonArray axes = multiscales.getAsJsonArray().get(0).getAsJsonObject().get("axes").getAsJsonArray();
+            int index = 0;
+            List<ZarrAxe> zarrAxes = new ArrayList<>();
```
Comment (tischi) [inline src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java:59]: @KateMoreva `ZarrAxe` is this supposed to be the singular of `ZarrAxes`?
If so, correct would be singular: `ZarrAxis`, plural: `ZarrAxes`.

File: src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java [hunk 2]
```diff
@@ -42,12 +44,41 @@ default String getDimensionSeparator(HashMap<String, JsonElement> attributes) {
...
+            for (JsonElement axe : axes) {
```
Comment (tischi) [inline src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java:60]: same here, probably you mean:
`axis : axes`

File: src/main/java/org/embl/mobie/io/ome/zarr/util/UnitTypes.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package org.embl.mobie.io.ome.zarr.util;
+
+public enum UnitTypes {
```
Comment (tischi) [inline src/main/java/org/embl/mobie/io/ome/zarr/util/UnitTypes.java:3]: @constantinpape @KateMoreva @joshmoore
Would it be better to depend on some other library for this?

**Comment by tischi** [comment]
@KateMoreva I had a few comments.

**Comment by KateMoreva** [comment]
> Do we have v0.4 example data for the demo?

Yes, we do. You can test them here, without MoBIE: src/test/java/ui/OmeZarrV4FSOpener.java, src/test/java/ui/OmeZarrS3V4Opener.java

File: src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java [hunk 4]
```diff
@@ -56,29 +55,29 @@ default void getDimensions(HashMap<String, JsonElement> attributes) throws Illeg
...
                 if (name.isEmpty() || type.isEmpty() || !AxesTypes.contains(type)) {
                     throw new IllegalArgumentException("Unsupported multiscales axes: " + name + ", " + type);
                 }
-                ZarrAxe zarrAxe;
+                ZarrAxis zarrAxe;
                 if (axe.getAsJsonObject().get("unit") != null && axe.getAsJsonObject().get("unit").isJsonPrimitive()) {
                     String unit = axe.getAsJsonObject().get("unit").getAsString();
                     if (UnitTypes.contains(unit)) {
-                        zarrAxe = new ZarrAxe(index, name, type, unit);
+                        zarrAxe = new ZarrAxis(index, name, type, unit);
```
Comment (tischi) [inline src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java:70]: also the variable name should be changed accordingly.

**Review by tischi — COMMENTED** [review]
please also have a look at the variable names.

File: src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java (hunk 1 above)
Comment (KateMoreva) [reply src/main/java/org/embl/mobie/io/ome/zarr/util/N5ZarrImageReader.java]: Yes, my mistake. Good point, thank you!

File: src/main/java/org/embl/mobie/io/ome/zarr/util/UnitTypes.java (hunk 3 above)
Comment (KateMoreva) [reply src/main/java/org/embl/mobie/io/ome/zarr/util/UnitTypes.java:3]: Yes, although I think there is nothing wrong with a double-check right now. But it is better to move it to `org.janelia.saalfeldlab.n5` once v4 becomes stable.

**Comment by KateMoreva** [comment]
> @KateMoreva I had a few comments.

@tischi, is everything else good?

**Comment by tischi** [comment]
Looks OK, I don't have time for a more detailed review.

**Comment by constantinpape** [comment]
@tischi do you think you could make a mobie-beta release with these changes still today?

**Comment by tischi** [comment]
yes, i can. but only after 17:00.

**Comment by constantinpape** [comment]
> yes, i can. but only after 17:00.

Great, just ping me when it's there :).

**Pull request merged**